src/mcpwm/generator.rs

```diff
@@ -0,0 +1,307 @@
+use core::{marker::PhantomData, ptr};
+
+use esp_idf_sys::{
+    esp, mcpwm_gen_handle_t, mcpwm_gen_timer_event_action_t, mcpwm_generator_action_t,
+    mcpwm_generator_action_t_MCPWM_GEN_ACTION_HIGH, mcpwm_generator_action_t_MCPWM_GEN_ACTION_KEEP,
+    mcpwm_generator_action_t_MCPWM_GEN_ACTION_LOW,
+    mcpwm_generator_action_t_MCPWM_GEN_ACTION_TOGGLE, mcpwm_generator_config_t,
+    mcpwm_generator_config_t__bindgen_ty_1, mcpwm_new_generator, mcpwm_oper_handle_t,
+    mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_DOWN,
+    mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_UP, mcpwm_timer_event_t_MCPWM_TIMER_EVENT_EMPTY,
+    mcpwm_timer_event_t_MCPWM_TIMER_EVENT_FULL, mcpwm_timer_event_t_MCPWM_TIMER_EVENT_INVALID,
+};
+
+use crate::gpio::OutputPin;
+
+use super::comparator::{Comparator, OptionalCmp};
+
+pub struct NoGen;
+
+impl OptionalGen for NoGen {}
+pub trait OptionalGen {}
+
+impl<G, CMPX, CMPY, P> OptionalGen for Generator<G, CMPX, CMPY, P>
+where
+    G: GeneratorChannel,
+    CMPX: OnMatchCfg,
+    CMPY: OnMatchCfg,
+    P: OutputPin,
+{
+}
+
+pub trait GeneratorChannel {
+    const IS_A: bool;
+}
+
+pub struct GenA;
+impl GeneratorChannel for GenA {
+    const IS_A: bool = true;
+}
+
+pub struct GenB;
+impl GeneratorChannel for GenB {
+    const IS_A: bool = false;
+}
+
+// TODO: Allow OptionalOutputPin?
+pub struct Generator<G, CMPX, CMPY, P: OutputPin> {
+    channel: PhantomData<G>,
+    cmp_x: PhantomData<CMPX>,
+    cmp_y: PhantomData<CMPY>,
+    pub(crate) _handle: mcpwm_gen_handle_t,
+    pub(crate) _pin: P,
+}
+
+pub struct GeneratorConfig<G: GeneratorChannel, CMPX, CMPY, P> {
+    _channel: PhantomData<G>,
+    pub(crate) flags: mcpwm_generator_config_t__bindgen_ty_1,
+    pub(crate) on_matches_cmp_x: CMPX,
+    pub(crate) on_matches_cmp_y: CMPY,
+    pub(crate) on_is_empty: CountingDirection,
+    pub(crate) on_is_full: CountingDirection,
+    pub(crate) pin: P,
+}
+
+pub struct NoGenCfg;
+
+pub trait OptionalGenCfg {
```

Comment:
Rather than introducing `OptionalGenCfg` trait, why not just using `Option<GeneratorConfig>`?